### PR TITLE
K8SPS-32: Orchestrator raft setup 

### PR DESCRIPTION
**Pull request by egegunes**
All nodes have their own database: sqlite. They communicate between each other by raft protocol.

All nodes run topology discoveries and failure detections but **only the leader runs failure recoveries**.

Multiple independent `orchestrator` nodes will each run failure analysis and there will be a **failure detection quorum**. If each `orchestrator` node runs in its own DC, the cluster will have a defense mechanism against network partitions:
* If one DC is partitioned away and the `orchestrator` in that DC is isolated, it can't be the leader.
* The other `orchestrator` nodes will agree on the type of failure with a quorum. They'll elect a new leader and the leader will run the recovery.

The list of raft nodes needs to be preconfigured:

```
  "RaftEnabled": true,
  "RaftDataDir": "/var/lib/orchestrator",
  "RaftAdvertise" "<ip.or.fqdn.of.this.orchestrator.node>",
  "RaftBind": "<ip.or.fqdn.of.this.orchestrator.node>",
  "DefaultRaftPort": 10008,
  "RaftNodes": [
    "<ip.or.fqdn.of.orchestrator.node1>",
    "<ip.or.fqdn.of.orchestrator.node2>",
    "<ip.or.fqdn.of.orchestrator.node3>"
  ],
```

Even if we use FQDNs in the configuration, the orchestrator resolves the IP address of the FQDN and uses it to bind/connect. This behavior makes using headless service FQDNs very fragile since the IP address of a pod can change during a rollout. For example the IP address of the pod `cluster1-orc-0` can change during the rollout but `cluster1-orc-1` and `cluster1-orc-2` still tries to connect to old IP.

To make the setup more resilient against the IP changes, we're creating a service per pod for `orchestrator` by default. The operator creates a `ConfigMap` with a list of FQDNs and mounts it to the `orchestrator` containers. `orchestrator` entrypoint uses this list as `RaftNodes` in
configuration.

Notice there are two fields: `RaftBind` and `RaftAdvertise`. In `RaftBind` we use the headless service FQDN, because each `orchestrator` node needs to bind to its IP address. `RaftAdvertise` is for other nodes to connect, so we're using the service FQDN.

We can’t rely on ConfigMap to scale Orchestrator up and down since it’ll trigger a new rollout restart on every scaling operation. The operator determines new and old peers by comparing the existing ConfigMap contents to the current state. New peers are added by “GET /api/raft-add-peer” to the leader and old ones are deleted by “GET /api/raft-remove-peer”. The operator always sends requests to the first pod (i.e. cluster1-orc-0) since the Orchestrator proxies the requests to the leader.

There are certain failure scenarios. For example scaling down to 1 pod may fail depending on who was the leader before operation. If the leader was the last pod for a 3 node cluster (i.e. cluster1-orc-2), the first pod is stuck in Candidate state. We force users to use a reliable setup for Orchestrator (unless they enable unsafe configurations): at least 3 nodes and the number of nodes should be an odd number.

Accompanying PR: https://github.com/percona/percona-docker/pull/557

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/5997a68/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/5997a68/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/5997a68/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/5997a68/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/5997a68/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/5997a68/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/5997a68/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/71/commits/5997a681222e25ef75081d7aac261ac42846fdfd
image: `perconalab/percona-server-mysql-operator:PR-71-5997a68`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/dd8aa3f/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/dd8aa3f/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/dd8aa3f/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/dd8aa3f/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/dd8aa3f/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/dd8aa3f/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/dd8aa3f/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/71/commits/dd8aa3f873d845d00359c28bf8e128f3d7264667
image: `perconalab/percona-server-mysql-operator:PR-71-dd8aa3f`

**Review by hors — CHANGES_REQUESTED** [review]
@egegunes can we scale up/down pods without full orchestrator cluster termination ?

Also I have scale down cluster then scale up it back and cluster can't choose the leader. 
In the logs I see (from all orc pods)
`2022/04/19 11:00:57 [INFO] I am not the leader. Sleeping ...`

But from orc log I can see that we have a leader 
`2022-04-19 11:16:44 DEBUG raft leader is 10.32.14.148:10008 (this host); state: Leader`
we need to improve [add_mysql_nodes.sh](https://github.com/percona/percona-docker/pull/557/files#diff-c58c0ac4b456d933371f8167948fdbbe9f170b1cbfc35d735325ec1aafa9cc4d)

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/6f6cb3f/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/6f6cb3f/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/6f6cb3f/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/6f6cb3f/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/6f6cb3f/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/6f6cb3f/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/6f6cb3f/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/71/commits/6f6cb3f07057bea225c7d63a13b70ad157426b6f
image: `perconalab/percona-server-mysql-operator:PR-71-6f6cb3f`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/2a8225e/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/2a8225e/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/2a8225e/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/2a8225e/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/2a8225e/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/2a8225e/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/2a8225e/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/71/commits/2a8225e9b7283fa0e387ef1f9d4e2634e51d1894
image: `perconalab/percona-server-mysql-operator:PR-71-2a8225e`

**Comment by egegunes** [comment]
@hors 

> @egegunes can we scale up/down pods without full orchestrator cluster termination ?

Scaling up was working but when you scale down to 1, the pod was terminating due to environment variable change. Now it's fixed. Also now we block scaling down to 1 unless users enable `allowUnsafeConfigurations`.

> 
> Also I have scale down cluster then scale up it back and cluster can't choose the leader. In the logs I see (from all orc pods) `2022/04/19 11:00:57 [INFO] I am not the leader. Sleeping ...`
> 
> But from orc log I can see that we have a leader `2022-04-19 11:16:44 DEBUG raft leader is 10.32.14.148:10008 (this host); state: Leader` we need to improve [add_mysql_nodes.sh](https://github.com/percona/percona-docker/pull/557/files#diff-c58c0ac4b456d933371f8167948fdbbe9f170b1cbfc35d735325ec1aafa9cc4d)

Please see https://github.com/percona/percona-docker/pull/557

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/59b3366/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/59b3366/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/59b3366/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/59b3366/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/59b3366/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/59b3366/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/59b3366/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/71/commits/59b3366370206b554050040d68b1ca94dac469a8
image: `perconalab/percona-server-mysql-operator:PR-71-59b3366`

**Review by pooknull — CHANGES_REQUESTED** [review]

**Review by tplavcic — DISMISSED** [review]

File: pkg/orchestrator/client.go [hunk 1]
```diff
@@ -29,61 +29,151 @@ type Instance struct {
 }
 
 func ClusterPrimary(ctx context.Context, apiHost, clusterHint string) (*Instance, error) {
+	url := fmt.Sprintf("%s/api/master/%s", apiHost, clusterHint)
+
+	resp, err := doRequest(ctx, url)
+	if err != nil {
+		return nil, errors.Wrapf(err, "do request to %s", url)
+	}
+	defer resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (egegunes) [reply pkg/orchestrator/client.go:38]: I think we do, why do you think that?

Quoting the go docs:

> If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close.

File: pkg/orchestrator/client.go [hunk 2]
```diff
@@ -29,61 +29,151 @@ type Instance struct {
...
 	}
 
+	if orcResp.Code == "ERROR" {
+		return errors.New(orcResp.Message)
+	}
+
 	return nil
 }
+
+func doRequest(ctx context.Context, url string) (*http.Response, error) {
```
[the remark this answers is not in the document]
Comment (egegunes) [reply pkg/orchestrator/client.go:168]: fixed in https://github.com/percona/percona-server-mysql-operator/pull/71/commits/05b08baf25c47a9ae27fd8ab5ae7a6f2979d7d4d

File: pkg/util/slice.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package util
+
+// Difference returns items that exist in b but not in a
+func Difference(a, b []string) (diff []string) {
+	m := make(map[string]bool)
```
[the remark this answers is not in the document]
Comment (egegunes) [reply pkg/util/slice.go]: fixed in https://github.com/percona/percona-server-mysql-operator/pull/71/commits/05b08baf25c47a9ae27fd8ab5ae7a6f2979d7d4d 

but I wonder what's the difference?

**Review by hors — CHANGES_REQUESTED** [review]

File: api/v1alpha1/perconaservermysql_types.go [hunk 4]
```diff
@@ -319,6 +323,10 @@ func (cr *PerconaServerMySQL) CheckNSetDefaults(serverVersion *platform.ServerVe
 	cr.Spec.MySQL.reconcileAffinityOpts()
 	cr.Spec.Orchestrator.reconcileAffinityOpts()
 
+	if oSize := int(cr.Spec.Orchestrator.Size); (oSize < 3 || oSize%2 == 0) && !cr.Spec.AllowUnsafeConfig {
+		return errors.New("Orchestrator size must be greater than 3 and an odd number for raft setup")
```
[the remark this answers is not in the document]
Comment (egegunes) [reply api/v1alpha1/perconaservermysql_types.go]: right :D https://github.com/percona/percona-server-mysql-operator/pull/71/commits/9ed1ad24e48e731dce8c1e26c3f99679ca32384f

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/05b08ba/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/05b08ba/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/05b08ba/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/05b08ba/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/05b08ba/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/05b08ba/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/05b08ba/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/71/commits/05b08baf25c47a9ae27fd8ab5ae7a6f2979d7d4d
image: `perconalab/percona-server-mysql-operator:PR-71-05b08ba`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/9ed1ad2/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/9ed1ad2/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/9ed1ad2/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/9ed1ad2/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/9ed1ad2/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/9ed1ad2/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/9ed1ad2/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/71/commits/9ed1ad24e48e731dce8c1e26c3f99679ca32384f
image: `perconalab/percona-server-mysql-operator:PR-71-9ed1ad2`

File: pkg/util/slice.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pooknull) [reply pkg/util/slice.go]: I think it is good practice to use empty structs if we are only checking for existing keys. This way, people who would read the code wouldn't be confused with the bool value

File: pkg/orchestrator/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pooknull) [reply pkg/orchestrator/client.go:38]: You are right. My bad

**Review by pooknull — DISMISSED** [review]

File: deploy/cr.yaml [hunk 5]
```diff
@@ -72,10 +72,23 @@ spec:
 #            storage: 1G
 
   orchestrator:
-    image: perconalab/percona-server-mysql-operator:main-orchestrator
+    image: perconalab/percona-server-mysql-operator:k8sps-32-orchestrator
```
Comment (hors) [inline deploy/cr.yaml:75]:
```suggestion
    image: perconalab/percona-server-mysql-operator:main-orchestrator
```

**Review by hors — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| config | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/edd6794/config.log) |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/edd6794/init-deploy.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/edd6794/monitoring.log) |
| semi-sync | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/edd6794/semi-sync.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/edd6794/service-per-pod.log) |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/edd6794/sidecars.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-ps-operator/PR-71/edd6794/users.log) |

commit: https://github.com/percona/percona-server-mysql-operator/pull/71/commits/edd6794b87711dc19d83b5d3cfd2d230c458901a
image: `perconalab/percona-server-mysql-operator:PR-71-edd6794`

**Review by hors — APPROVED** [review]

**Review by tplavcic — APPROVED** [review]

**Pull request merged**